### DISCUSSION TITLE
How to convert from dosemu to dosemu2

### DISCUSSION THREAD

**Discussion by nelgin** [dosemu2/dosemu2#1651] [General]
I'm running a BBS which, using dosemu, allows dos doors to be run on a linux platform by utilizing a number of scripts. One of the basis is that an external.bat file is created in the user's node directory which is run when dosemu.bin is invoked.

```
HOME=/sbbs/ctrl/ QUIET=1 DOSDRIVE_D=/sbbs/node2/ NODEDIR=/sbbs/node2/ /usr/bin/dosemu.bin -I"video { none }" -I"serial { virtual com 1 }" -f/etc/dosemu/dosemu.conf -ED:external.bat -o/sbbs/node2/dosemu_boot.log
```

This is the command which is run. /sbbs/ctrl is where the .dosemu directory is while /sbbs/node2 (or /sbbs/nodex) is where the external.bat will be placed.

Since DOSDRIVE_D has been eliminated, I'm not sure how to replicate this setup so it'll work in dosemu2.

external.bat does a number of drive redirections:

```
SET DSZLOG=D:\PROTOCOL.LOG
SET SBBSNODE=D:
SET SBBSNNUM=2
SET SBBSCTRL=F:
SET SBBSDATA=G:
SET SBBSEXEC=H:
SET PCBNODE=2
SET PCBDRIVE=D:
SET PCBDIR=\
SET STARTDIR=dbt
lredir D: linux\fs\sbbs\node2 >NUL
lredir E: linux\fs\sbbs\xtrn >NUL
lredir F: linux\fs\sbbs\ctrl >NUL
lredir G: linux\fs\sbbs\data >NUL
lredir H: linux\fs\sbbs\exec >NUL

E:
```
from there it does off to run the game. 

Simple question, I hope, how can I fix this to work with dosemu2 ?

**Comment by stsp** [comment]
Just a basic suggestions.

Try to avoid `-I` stuff. Use
options in `~/.dosemu/.dosemurc`
instead, or the cmdline options
like `-dumb` instead of video none,
$_comX instead of `-I serial` etc.

Use $_hostfs_drives option instead
of custom redirections in a bat file.

Use `VAR=val` syntax instead of
setting things in a bat file, eg
`dosemu SBBSNNUM=2`
will pass that var to DOS.

dosemu2 may occupy some drive
letters, and you should either move
your stuff to other drive letters, or
adjust $_hdimage option to move
dosemu2 drives to other letters.
See comments in dosemu.conf for
that option - there are many examples
provided.

Sorry but you took the most unportable
and evil ways of configuring dosemu1
(using -I is very evil and unportable),
so the migration will take some efforts.
But if you follow the above suggestions,
your setup will likely not break again.

**Reply by stsp** [reply]
> Since DOSDRIVE_D has been eliminated, I'm not sure how to replicate this setup so it'll work in dosemu2.

Also pay attention to -d
option that allows you to
add new drives, somewhat
similar to DOSDRIVE_D.

**Reply by nelgin** [reply]
Thank you for this valuable insight. This was a legacy implementation in the BBS package I'm using so had no control over how it was added. I can just hope to fix it.

**Comment by runderwo** [comment]
The original Synchronet dosemu glue was written 20 years ago and is completely my fault.  Try to work around the DOSDRIVE_D issue first with the CLI option Stas posted, and get some non-problematic door like TW2002 or LORD working.  Then once you're back to a working place, pursue the other cleanups.  The reason I suggest doing it this way is that it's very tricky to get things just right so that doors launch correctly, and incremental changes will be less frustrating than trying to debug a sweeping change.

**Comment by nelgin** [comment]
Thanks have changed a bit since it was implemented. Since the big sticking point is the DOSDRIVE_D thing, why every they removed it I have no idea. It worked and wouldn't have broke things that still use it. Oh well. It'd be nice if they documented what to do instead rather than saying that it's now DOSDRIVE_EXTRA which doesn't work either and I find no trace of in the source code.  I'll work on it.

**Reply by nelgin** [reply]
Success!

https://pasteboard.co/SndM9lDhGwnY.png

**Reply by stsp** [reply]
> Since the big sticking point is the DOSDRIVE_D thing, why every they removed it I have no idea.

dosemu2 uses some drives
internally, including D by default.
So some things were removed,
but not the functionality itself.
Same things are still possible.

> saying that it's now DOSDRIVE_EXTRA which doesn't work either

Who said about DOSDRIVE_EXTRA?
It is documented anywhere?
Where have you found that artefact?

> Success!

Cool!
Where is the end result? :)

**Reply by nelgin** [reply]
> Who said about DOSDRIVE_EXTRA?
> It is documented anywhere?
> Where have you found that artefact?

https://github.com/dosemu2/dosemu2/releases

"DOSDRIVE_D var is removed. Use new DOSDRIVE_EXTRA variable."

So yeah, if you're going to have it in the release notes, you might want to add it to the program, or remove from the release notes, one or the other.

As for the end result. it failed to do what I needed in the end, so rolled it back to dosemu1 to stay compatible with everyone else using the same BBS software.

**Comment by andrewbird** [comment]
> So yeah, if you're going to have it in the release notes, you might want to add it to the program, or remove from the release notes, one or the other.

Mmm, difficult one that. It's not really desirable to rewrite release notes from 6 years ago if they are not longer accurate. I suppose a subsequent release note might have mentioned the variable's demise.

In general documentation for Dosemu2 is difficult, there are so many snippets of information floating around the Internet for the original Dosemu from a time when that was very popular. They may or may not still be relevant as some things have had to change, kernel support disappearing, 32bit distros no longer available, security concerns have had to be addressed, and some different goals for the project too. The project wiki on github is open to all and has a few pages that people have written to help each other.

> As for the end result. it failed to do what I needed in the end, so rolled it back to dosemu1 to stay compatible with everyone else using the same BBS software.

So that was the success you mentioned, or did you get Dosemu2 working first and found a limitation?

**Comment by nelgin** [comment]
I was trying to get some DOS BBS doors working. Trade Wars was a notorious problem in dosemu and people had got it working under dosemu2 but with very few instructions. The main program to create the "universe" is BIGBANG which would just kill dosemu without much of a meaningful error (to me anyway). I got it working in dosemu2 without too much of a problem, but then I tried to run another and I had some issue and decided I had spent too many hours on it already. As you say, it's probably down to some kernel issue. I'm running 5.13.0 currently and have no intentions of recompiling to get anything working. 

All the work I did was on a test system that I blew away when I was finished with it.  I may revisit one day.

**Reply by jschwartzenberg** [reply]
Note that there's a real strive to improve DOSEMU2 but it can be hard when there is little information on a particular problem. If you can provide logs, instructions to reproduce, etc. it will help to find a solution.

**Comment by andrewbird** [comment]
> As you say, it's probably down to some kernel issue.

Not quite what I meant. Changes have been made to modern kernels for good reason I believe, but that can mean that things Dosemu1 used to do aren't possible any more. Dosemu2 has had to change the way it does some things in order to work with those changes. I wouldn't expect that any user of Dosemu2 would have to recompile the kernel to make it work.

> All the work I did was on a test system that I blew away when I was finished with it. I may revisit one day.

That's a shame. If you do revisit some day, please open an issue here for each problem as you find them, following the template so that you collect all the logs, configs and other info so that we may figure out what's going on.

**Reply by nelgin** [reply]
@andrewbird Challenge accepted. I have setup a new environment on my local system. but if I'm going to spend my time setting up this environment and testing doors and reporting the bugs, I hope someone will be serious and dedicate time to resolving any issues. I'm happy to try pretty much whatever.

**Comment by stsp** [comment]
Well, you've found DOSDRIVE_EXTRA
in some ancient release notes, _after_ I
told you that you can use -d.
So there was no need of such searching.

I am not sure how bad is that you blew
that system, at least wrt dosemu. Maybe
it took a lot of time of you to find some
solutions, but since all these solutions
are just a few switches and config options,
I don't believe it will take more than 10
minutes to get to where you stopped,
as you probably won't search for them
again but reproduce from memory.
